pkg/execution/debounce/debounce.go

```diff
@@ -0,0 +1,362 @@
+	}
+
+	if out == "0" {
+		// Enqueue the debounce job with extra buffer *plus* one second.  This ensures that we never
+		// attempt to start a debounce during the debounce's expiry (race conditions), and the extra
+		// second lets an updateDebounce call on TTL 0 finish, as the buffer is the updateDebounce
+		// deadline.
+		qi := d.queueItem(ctx, di, debounceID)
+		err = d.q.Enqueue(ctx, qi, now.Add(ttl).Add(buffer).Add(time.Second))
+		if err != nil {
+			return &debounceID, fmt.Errorf("error enqueueing debounce job: %w", err)
+		}
+		return &debounceID, nil
+	}
+
+	existingID, err := ulid.Parse(out)
+	if err != nil {
+		// This was not a ULID, so we have no idea what was returned.
+		return nil, fmt.Errorf("unknown new debounce return value: %s", out)
+	}
+	return &existingID, ErrDebounceExists
+}
+
+// updateDebounce updates the currently pending debounce to point to the new event ID.  It pushes
+// out the debounce's TTL, and re-enqueues the job to initialize fns from the debounce.
+func (d debouncer) updateDebounce(ctx context.Context, di DebounceItem, fn inngest.Function, ttl time.Duration, debounceID ulid.ULID) error {
+	now := time.Now()
+
+	key, err := d.debounceKey(ctx, di, fn)
+	if err != nil {
+		return err
+	}
+
+	// NOTE: This function has a deadline to complete.  If this fn doesn't complete within the deadline,
+	// eg, network issues, we must check if the debounce expired and re-attempt the entire thing, allowing
+	// us to either update or create a new debounce depending on the current time.
+	ctx, cancel := context.WithTimeout(ctx, buffer)
+	defer cancel()
+
+	keyPtr := d.k.DebouncePointer(ctx, fn.ID, key)
+	keyDbc := d.k.Debounce(ctx)
+	byt, err := json.Marshal(di)
+	if err != nil {
+		return fmt.Errorf("error marshalling debounce: %w", err)
+	}
+
+	out, err := scripts["updateDebounce"].Exec(
+		ctx,
+		d.r,
+		[]string{keyPtr, keyDbc},
+		[]string{debounceID.String(), string(byt), strconv.Itoa(int(ttl.Seconds()))},
+	).AsInt64()
+	if err != nil {
+		return fmt.Errorf("error creating debounce: %w", err)
+	}
+	switch out {
+	case 0:
+		err = d.q.RequeueByJobID(
+			ctx,
+			fn.ID.String(),
+			debounceID.String(),
+			now.Add(ttl).Add(buffer).Add(time.Second),
+		)
+		if err != nil {
+			return fmt.Errorf("error requeueing debounce job '%s': %w", debounceID, err)
+		}
+		return nil
+	default:
+		return fmt.Errorf("unknown update debounce return value: %d", out)
+	}
+}
+
+func (d debouncer) debounceKey(ctx context.Context, evt event.TrackedEvent, fn inngest.Function) (string, error) {
+	if fn.Debounce.Key == nil {
+		return fn.ID.String(), nil
+	}
+
+	out, _, err := expressions.Evaluate(ctx, *fn.Debounce.Key, map[string]any{"event": evt.GetEvent().Map()})
+	if err != nil {
+		return "", fmt.Errorf("invalid debounce expression: %w", err)
+	}
+	if str, ok := out.(string); ok {
+		return str, nil
+	}
+	return fmt.Sprintf("%v", out), nil
+}
+
+func readRedisScripts(path string, entries []fs.DirEntry) {
```

Comment:
non-blocking:

this probably should be a utility function or something.
there's probably like 3 copies of this code in the entire code base at this point.